# My Coursera Capstone Project

This Jupyter Notebook will be used to show my work for my Coursera Capston project for the IBM Data Science Professional Certificate.

In [61]:
import pandas as pd
import numpy as np

In [3]:
print('Hello Capstone Project Course!')

Hello Capstone Project Course!


#### Question No. 1 Segmenting and Clustering Neighborhoods in Toronto: Downloading and Webscrapping

##### Installing packages and Downloading the data

In [80]:
!pip install bs4
from bs4 import BeautifulSoup
import requests
import pandas as pd
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
print('import complete')

import complete


In [87]:
#Prepping the Data for Webscrapping 
data = requests.get(url).text
soup = BeautifulSoup(data, "html5lib")
table = soup.find('table')

In [91]:
#Webscrapping 
table_contents =[]
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace('/',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)


In [92]:
# Creating the Pandas Dataframe (df)
df=pd.DataFrame(table_contents)

df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                     'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                     'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                     'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

In [98]:
df.head()

PostalCode           Borough                       Neighborhood
0        M3A        North York                          Parkwoods
1        M4A        North York                   Victoria Village
2        M5A  Downtown Toronto         Regent Park , Harbourfront
3        M6A        North York  Lawrence Manor , Lawrence Heights
4        M7A      Queen's Park      Ontario Provincial Government

In [94]:
df.shape

(103, 3)

#### Question No. 2 of the Assignment: Latitude and Longitude

In [105]:
geodf=pd.read_csv(r'C:\Users\User\Desktop\Coursera\IBM Data Science Certificate\Course 10 Applied Data Science Capstone\Geospatial_Coordinates.csv')
geodf.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [106]:
#Renaming the Postal Code Column to be able to merge
geodf.rename(columns={'Postal Code': 'PostalCode'}, inplace=True)
geodf.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

##### The New Merged Dataframe

In [108]:
Torontodf = pd.merge(df, geodf, on='PostalCode', how='inner')
Torontodf.head()

PostalCode           Borough                       Neighborhood   Latitude  \
0        M3A        North York                          Parkwoods  43.753259   
1        M4A        North York                   Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park , Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor , Lawrence Heights  43.718518   
4        M7A      Queen's Park      Ontario Provincial Government  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

#### Question No. 3 - Exploring the Neighborhood

In [115]:
import json
!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim
import requests
from pandas.io.json import json_normalize 
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
!conda install -c conda-forge folium=0.5.0 --yes 
import folium 



==> WARNING: A newer version of conda exists. <==
  current version: 4.10.1
  latest version: 4.10.3

Please update conda by running

    $ conda update -n base -c defaults conda




Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\User\anaconda3

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    conda-4.10.3               |   py38haa244fe_0         3.1 MB  conda-forge
    geographiclib-1.52         |     pyhd8ed1ab_0          35 KB  conda-forge
    geopy-2.2.0                |     pyhd8ed1ab_0          67 KB  conda-forge
    python_abi-3.8             |           2_cp38           4 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         3.2 MB

The following NEW packages will be INSTALLED:

  geographiclib      conda-forge/noarch::geographiclib-1.52-pyhd8ed1ab_0
  geopy              conda-forge/noarch::geopy-2.2.0-pyhd8ed1ab_0
  python_abi         conda-forge/win-64::python_abi-3.8-2_cp38

The following 

In [111]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(Torontodf['Borough'].unique()),
        Torontodf.shape[0]
    )
)

The dataframe has 15 boroughs and 103 neighborhoods.


In [116]:
#Finding the Toronto Latitude and Longitude 
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [121]:
#Creating a map of Toronto, ON using Lat, Long
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)


for lat, lng, borough, neighborhood in zip(Torontodf['Latitude'], Torontodf['Longitude'], Torontodf['Borough'], Torontodf['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

In [123]:
CLIENT_ID = '3QCOYLJIYELE1O32XIQQFCQOA1PEHSTZCM0POEIJMUZUR10A' 
CLIENT_SECRET = 'P1AC0GGYPAPZF4JJOS03JHWJ4PMY43FQAVMQCL25BTM5ZPOT' 
VERSION = '20180605'
LIMIT = 100 

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 3QCOYLJIYELE1O32XIQQFCQOA1PEHSTZCM0POEIJMUZUR10A
CLIENT_SECRET:P1AC0GGYPAPZF4JJOS03JHWJ4PMY43FQAVMQCL25BTM5ZPOT


In [124]:
Torontodf.loc[4, 'Neighborhood']

'Ontario Provincial Government'

In [125]:
neighborhood_latitude = Torontodf.loc[4, 'Latitude']
neighborhood_longitude = Torontodf.loc[4, 'Longitude']

neighborhood_name = Torontodf.loc[4, 'Neighborhood']

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Ontario Provincial Government are 43.6623015, -79.3894938.


#### Top 50 Venues Around Ontario Provincial Government

In [126]:
LIMIT = 50
radius = 500

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=3QCOYLJIYELE1O32XIQQFCQOA1PEHSTZCM0POEIJMUZUR10A&client_secret=P1AC0GGYPAPZF4JJOS03JHWJ4PMY43FQAVMQCL25BTM5ZPOT&v=20180605&ll=43.6623015,-79.3894938&radius=500&limit=50'

In [127]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '60facba0a566972081fb40cb'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Bay Street Corridor',
  'headerFullLocation': 'Bay Street Corridor, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 30,
  'suggestedBounds': {'ne': {'lat': 43.6668015045, 'lng': -79.38328496866619},
   'sw': {'lat': 43.657801495499996, 'lng': -79.3957026313338}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b9d206bf964a520e69136e3',
       'name': "Queen's Park",
       'location': {'address': 'University Ave.',
        'crossStreet': 'at Wellesley Ave.',
        'lat': 43.66394609897775,
        'lng': -79.39217952520835,
        'labeledL

In [128]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [132]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

<ipython-input-132-561c05f0fdd1>:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  nearby_venues = json_normalize(venues) # flatten JSON


name          categories        lat        lng
0    Queen's Park                Park  43.663946 -79.392180
1        Mercatto  Italian Restaurant  43.660391 -79.387664
2  NEO COFFEE BAR         Coffee Shop  43.660130 -79.385830
3        Bar Volo            Beer Bar  43.665462 -79.385692
4   T-Swirl Crepe            Creperie  43.663452 -79.384125

In [130]:
#Venues Returned by Foursquare
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

30 venues were returned by Foursquare.


### Question No. 3b Exploring Neighborhoods in Toronto

In [133]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [135]:
Toronto_venues = getNearbyVenues(names=Torontodf['Neighborhood'],
                                   latitudes=Torontodf['Latitude'],
                                   longitudes=Torontodf['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park , Harbourfront
Lawrence Manor , Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern , Rouge
Don Mills North
Parkview Hill , Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park , Princess Gardens , Martin Grove , Islington , Cloverdale
Rouge Hill , Port Union , Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate , Bloordale Gardens , Old Burnhamthorpe , Markland Wood
Guildwood , Morningside , West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor , Wilson Heights , Downsview North
Thorncliffe Park
Richmond , Adelaide , King
Dufferin , Dovercourt Village
Scarborough Village
Fairview , Henry Farm , Oriole
Northwood Park , York University
The Danforth  East
Harbourfront East , Union Station , Toronto Islands
Little Portugal , Trinity
Kennedy Park , Ionview , East Birchmount Park
Bayview Village

In [136]:
print(Toronto_venues.shape)
Toronto_venues.head()

(1675, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1                     KFC       43.754387       -79.333021   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4             Tim Hortons       43.725517       -79.313103   

         Venue Category  
0                  Park  
1  Fast Food Restaurant  
2     Food & Drink Shop  
3          Hockey Arena  
4           Coffee Shop

In [137]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                               
Agincourt                                                              4   
Alderwood , Long Branch                                                8   
Bathurst Manor , Wilson Heights , Downsview North                     23   
Bayview Village                                                        4   
Bedford Park , Lawrence Manor East                                    24   
...                                                                  ...   
Willowdale South                                                      33   
Willowdale West                                                        6   
Woburn                                                                 3   
Woodbine Heights                                                       8   
York Mills West                                                        2   

                                                   Neighborhood Longitude  \
Neighborhood                                                                
Agincourt                                                               4   
Alderwood , Long Branch                                                 8   
Bathurst Manor , Wilson Heights , Downsview North                      23   
Bayview Village                                                         4   
Bedford Park , Lawrence Manor East                                     24   
...                                                                   ...   
Willowdale South                                                       33   
Willowdale West                                                         6   
Woburn                                                                  3   
Woodbine Heights                                                        8   
York Mills West                                                         2   

                                                   Venue  Venue Latitude  \
Neighborhood                                                               
Agincourt                                              4               4   
Alderwood , Long Branch                                8               8   
Bathurst Manor , Wilson Heights , Downsview North     23              23   
Bayview Village                                        4               4   
Bedford Park , Lawrence Manor East                    24              24   
...                                                  ...             ...   
Willowdale South                                      33              33   
Willowdale West                                        6               6   
Woburn                                                 3               3   
Woodbine Heights                                       8               8   
York Mills West                                        2               2   

                                                   Venue Longitude  \
Neighborhood                                                         
Agincourt                                                        4   
Alderwood , Long Branch                                          8   
Bathurst Manor , Wilson Heights , Downsview North               23   
Bayview Village                                                  4   
Bedford Park , Lawrence Manor East                              24   
...                                                            ...   
Willowdale South                                                33   
Willowdale West                                                  6   
Woburn                                                           3   
Woodbine Heights                                                 8   
York Mills West                                                  2   

                                                   Venue Category  
Neighborhood                                                       
Agincourt               

In [138]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 254 uniques categories.


##### There are a large amount of venues in Toronto; however, there are 254 unique venues in the neighborhoods of Toronto.

### Question No. 4 - Analyzing each Toronto Neighborhood

In [139]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Airport  \
0            0                  0               0        0   
1            0                  0               0        0   
2            0                  0               0        0   
3            0                  0               0        0   
4            0                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Truck Stop  \
0                 0                    0  ...           0   
1                 0                    0  ...           0   
2                 0                    0  ...           0   
3                 0                    0  ...           0   
4                 0                    0  ...           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                      0                0         0          0            0   
1                      0                0         0          0            0   
2                      0                0         0          0            0   
3                      0                0         0          0            0   
4                      0                0         0          0            0   

   Women's Store  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 254 columns]

In [140]:
Toronto_onehot.shape

(1675, 254)

###### Grouping neighborhoods and taking the mean of the frequency of the occurence of each category

In [141]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Toronto_grouped

Neighborhood  Yoga Studio  \
0                                           Agincourt          0.0   
1                             Alderwood , Long Branch          0.0   
2   Bathurst Manor , Wilson Heights , Downsview North          0.0   
3                                     Bayview Village          0.0   
4                  Bedford Park , Lawrence Manor East          0.0   
..                                                ...          ...   
95                                   Willowdale South          0.0   
96                                    Willowdale West          0.0   
97                                             Woburn          0.0   
98                                   Woodbine Heights          0.0   
99                                    York Mills West          0.0   

    Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                 0.0             0.0      0.0                 0.0   
1                 0.0             0.0      0.0                 0.0   
2                 0.0             0.0      0.0                 0.0   
3                 0.0             0.0      0.0                 0.0   
4                 0.0             0.0      0.0                 0.0   
..                ...             ...      ...                 ...   
95                0.0             0.0      0.0                 0.0   
96                0.0             0.0      0.0                 0.0   
97                0.0             0.0      0.0                 0.0   
98                0.0             0.0      0.0                 0.0   
99                0.0             0.0      0.0                 0.0   

    Airport Gate  Airport Lounge  Airport Service  Airport Terminal  ...  \
0            0.0             0.0              0.0               0.0  ...   
1            0.0             0.0              0.0               0.0  ...   
2            0.0             0.0              0.0               0.0  ...   
3            0.0             0.0              0.0               0.0  ...   
4            0.0             0.0              0.0               0.0  ...   
..           ...             ...              ...               ...  ...   
95           0.0             0.0              0.0               0.0  ...   
96           0.0             0.0              0.0               0.0  ...   
97           0.0             0.0              0.0               0.0  ...   
98           0.0             0.0              0.0               0.0  ...   
99           0.0             0.0              0.0               0.0  ...   

    Truck Stop  Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0          0.0                            0.0               0.0          0.0   
1          0.0                            0.0               0.0          0.0   
2          0.0                            0.0               0.0          0.0   
3          0.0                            0.0               0.0          0.0   
4          0.0                            0.0               0.0          0.0   
..         ...                            ...               ...          ...   
95         0.0                            0.0               0.0          0.0   
96         0.0                            0.0               0.0          0.0   
97         0.0                            0.0               0.0          0.0   
98         0.0                            0.0               0.0          0.0   
99         0.0                            0.0               0.0          0.0   

    Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                0.000000              0.0       0.0        0.0          0.0   
1                0.000000              0.0       0.0        0.0          0.0   
2                0.000000              0.0       0.0        0.0          0.0   
3                0.000000              0.0       0.0        0.0          0.0   
4                0.000000              0.0       0.0        0.0          0.0   
..   

In [142]:
Toronto_grouped.shape

(100, 254)

##### Top 5 Common Venues (As the above grouping did not show us much frequency)

In [143]:
num_top_venues = 5

for hood in Toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0  Latin American Restaurant  0.25
1                     Lounge  0.25
2             Breakfast Spot  0.25
3             Clothing Store  0.25
4  Middle Eastern Restaurant  0.00


----Alderwood , Long Branch----
         venue  freq
0  Pizza Place  0.25
1     Pharmacy  0.12
2         Pool  0.12
3  Coffee Shop  0.12
4          Pub  0.12


----Bathurst Manor , Wilson Heights , Downsview North----
                venue  freq
0                Bank  0.09
1         Coffee Shop  0.09
2  Frozen Yogurt Shop  0.04
3   Mobile Phone Shop  0.04
4    Sushi Restaurant  0.04


----Bayview Village----
                 venue  freq
0  Japanese Restaurant  0.25
1   Chinese Restaurant  0.25
2                 Bank  0.25
3                 Café  0.25
4          Yoga Studio  0.00


----Bedford Park , Lawrence Manor East----
                venue  freq
0         Coffee Shop  0.08
1      Sandwich Place  0.08
2  Italian Restaurant  0.08
3    Greek Restaurant  0.04

In [144]:
# Put into a pandas dataframe
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Now, let's see the top 10 venues for each neighborhood

In [146]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  \
0                                          Agincourt   
1                            Alderwood , Long Branch   
2  Bathurst Manor , Wilson Heights , Downsview North   
3                                    Bayview Village   
4                 Bedford Park , Lawrence Manor East   

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  Latin American Restaurant                Lounge        Breakfast Spot   
1                Pizza Place              Pharmacy                  Pool   
2                       Bank           Coffee Shop    Frozen Yogurt Shop   
3        Japanese Restaurant    Chinese Restaurant                  Bank   
4                Coffee Shop        Sandwich Place    Italian Restaurant   

  4th Most Common Venue      5th Most Common Venue  \
0        Clothing Store  Middle Eastern Restaurant   
1           Coffee Shop                        Pub   
2     Mobile Phone Shop           Sushi Restaurant   
3                  Café                Yoga Studio   
4      Greek Restaurant            Thai Restaurant   

        6th Most Common Venue            7th Most Common Venue  \
0         Monument / Landmark  Molecular Gastronomy Restaurant   
1              Sandwich Place                              Gym   
2                       Diner                       Restaurant   
3  Modern European Restaurant              Moroccan Restaurant   
4                 Pizza Place          Comfort Food Restaurant   

        8th Most Common Venue            9th Most Common Venue  \
0  Modern European Restaurant                Mobile Phone Shop   
1               Historic Site                           Museum   
2         Fried Chicken Joint                      Bridal Shop   
3         Monument / Landmark  Molecular Gastronomy Restaurant   
4                Cupcake Shop                              Pub   

  10th Most Common Venue  
0     Miscellaneous Shop  
1     Mexican Restaurant  
2         Sandwich Place  
3      Mobile Phone Shop  
4             Restaurant

### Question No. 4 - Clusting Neighborhoods Using K-Means

In [187]:
# set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 0])

In [188]:
# add clustering labels
Toronto_merged = Torontodf

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Toronto_merged.head()
Toronto_merged['Cluster Labels']=Toronto_merged['Cluster Labels'].fillna(0).astype('int') 

#### Visualize the map of Toronto with 5 Clusters that also reflect the top 10 venues for each neighborhood

In [189]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters